core/src/main/java/me/lokka30/treasury/plugin/core/utils/downloader/PluginDownloader.java

```diff
@@ -0,0 +1,215 @@
+            try (Reader in = new InputStreamReader(latestBuild.openStream())) {
+                JsonObject object = Utils.GSON.fromJson(in, JsonObject.class);
+                JsonObject build = object.getAsJsonArray("builds").get(0).getAsJsonObject();
+
+                // check dates
+                OffsetDateTime currentJarDate = OffsetDateTime.parse(PluginDownloader.class
+                        .getPackage()
+                        .getImplementationVersion());
+                OffsetDateTime buildDate = parseDate(build
+                        .getAsJsonObject("changeSet")
+                        .getAsJsonArray("items")
+                        .get(0)
+                        .getAsJsonObject()
+                        .get("date")
+                        .getAsString());
+                if (currentJarDate.isEqual(buildDate) || currentJarDate.isAfter(buildDate)) {
+                    if (source != null) {
+                        source.sendMessage(Message.of(MessageKey.DOWNLOAD_LATEST_ALREADY_LATEST));
+                    } else {
+                        plugin.logger().warn("Already running latest.");
+                    }
+                    return false;
+                }
+
+                String downloadUrlString = downloadPlatform.downloadBase();
+                downloadUrlString = downloadUrlString.replace("%number%",
+                        Integer.toString(build.get("number").getAsInt())
+                );
+
+                String downloadPath = null;
+                for (JsonElement element : build.getAsJsonArray("artifacts")) {
+                    JsonObject artifact = element.getAsJsonObject();
+                    String relativePath = artifact.get("relativePath").getAsString();
+                    if (relativePath.contains(platformName) && !relativePath.contains("-sources")) {
+                        downloadPath = relativePath;
+                        break;
+                    }
+                }
+                if (downloadPath == null) {
+                    // nag the user to nag us that something went wrong
+                    if (source != null) {
+                        source.sendMessage(Message.of(MessageKey.DOWNLOAD_LATEST_COULDNT_PARSE));
+                    } else {
+                        plugin.logger().warn(
+                                "Something went wrong whilst parsing latest build download data. Please notify a Treasury developer ASAP.");
+                    }
+                    return false;
+                }
+                downloadUrlString = downloadUrlString.concat(downloadPath);
+
+                // find old jar because after we download we may not be able to find it
+                File oldJar = null;
+                for (File file : plugin
+                        .pluginsFolder()
+                        .toFile()
+                        .listFiles((dir, name) -> name.contains("treasury"))) {
```

Comment:
> > `new File(MyClass.class.getProtectionDomain().getCodeSource().getLocation().toURI())`
> 
> What will this give us?

This gets the location of the class' CodeSource, which is (usually) the location of the jar file.

> > For Bukkit specifically
> 
> We shouldn't use a platform specific thing in my opinion in order to get the old jar.

This would be an alternative to `TreasuryPlugin#pluginsFolder`.

Reply:
> This would be an alternative to `TreasuryPlugin#pluginsFolder`.

We still need TreasuryPlugin#pluginsFolder because we have to create the new file. 
I'm gonna perform some tests and check what I can replace that code with

Reply:
> We still need `TreasuryPlugin#pluginsFolder` because we have to create the new file.
I'm gonna perform some tests and check what I can replace that code with

`File#getParent` would work - one assumes if it's running from a folder, the parent is the correct download location.

Reply:
I'm gonna do what's called a pro gamer move - use that code you provided but also have a method that could be or could not be implemented on a platform just in case to check the file we received from that code you provided. If that method is not implemented we assume it is the correct jar and delete it on exit. A simple test concluded the code you provided works fine on my instance, however we should be defensive.